### PR TITLE
[WIP] Prototype / hack of integration of DTS into sanitycheck

### PR DESCRIPTION
**Pull request by galak**
This is a wip in progress, first cut/example of using DTS in sanitycheck.  Looking for feedback if this direction seems desirable.  etc.

### CODE REVIEW

**Review by ulfalizer — COMMENTED** [review]
Haven't looked much at these scripts, but getting rid of `generated_dts_board.conf` seems nice.

File: scripts/sanity_chk/expr_parser.py [hunk 1]
```diff
@@ -10,6 +10,11 @@
 import threading
 import re
 
+ZEPHYR_BASE = os.environ.get("ZEPHYR_BASE")
+sys.path.insert(0, os.path.join(ZEPHYR_BASE, "scripts/dts"))
```
Comment (pabigot) [inline scripts/sanity_chk/expr_parser.py]: IIRC for portability this should be `os.path.join(ZEPHYR_BASE, "scripts", "dts")`  but there's probably already a lot of code that won't work for `os.path.sep` not equal to `'/'`.  Certainly there's material in `sanitycheck` that should be using `os.path.split` if that's a concern.

**Review by pabigot — APPROVED** [review]
Seems like a clean solution, especially if this means we can put application-specific board conf/overlay files in the test and they get picked up automatically (no mods to `testcase.yaml` or `sample.yaml`).

File: doc/guides/test/sanitycheck.rst [hunk 2]
```diff
@@ -548,15 +439,81 @@ Beside being able to run tests in QEMU and other simulated environments,
 sanitycheck supports running most of the tests on real devices and produces
 reports for each run with detailed FAIL/PASS results.
 
-To use this feature, run sanitycheck with the following new options::
+
+Executing tests on a single device
+===================================
+
+To use this feature on a single connected device, run sanitycheck with
+the following new options::
```
Comment (dbkinder) [inline doc/guides/test/sanitycheck.rst]: they won't be "new" for long, so just say "the following options"

File: doc/guides/test/sanitycheck.rst [hunk 3]
```diff
@@ -548,15 +439,81 @@ Beside being able to run tests in QEMU and other simulated environments,
...
 This needs to be accessible by the user running sanitycheck. You can run this on
-only one board at a time, specified using the
-``--platform`` option.
+only one board at a time, specified using the ``--platform`` option.
+
+
+Executing tests on multiple devices
+===================================
+
+To build and execute tests on multiple devices connected to the host PC a
```
Comment (dbkinder) [inline doc/guides/test/sanitycheck.rst]: Better would be:
```
To build and execute tests on multiple devices connected to the host PC,
you'll need to create a hardware map with all connected devices and their
details, such as the serial device and their IDs, if available. Create a
hardware map by running this command::
```

File: doc/guides/test/sanitycheck.rst [hunk 4]
```diff
@@ -548,15 +439,81 @@ Beside being able to run tests in QEMU and other simulated environments,
...
+    serial: /dev/cu.usbmodem146114202
+  - available: true
+    id: 000683759358
+    platform: unknown
+    product: J-Link
+    runner: unknown
+    serial: /dev/cu.usbmodem0006837593581
+
+
+Any options marked as 'unknown' need to be changed and set with the correct
```
Comment (dbkinder) [inline doc/guides/test/sanitycheck.rst]: Better would be:
```
You must manually change any options marked as ``unknown``.
In the above example, replace the ``platform`` and ``runner``
names with values appropriate for the connected
hardware. In this example we are using a ``reel_board`` and an
``nrf52840_pca10056`` board, and ``pyocd`` and ``nrfjprog``
runners, respectively::
```

File: doc/guides/test/sanitycheck.rst [hunk 5]
```diff
@@ -548,15 +439,81 @@ Beside being able to run tests in QEMU and other simulated environments,
...
+
+  ./scripts/sanitycheck --device-testing --hardware-map map.yml -T samples/hello_world/
+
+The above command will result in sanitycheck building tests for the platforms
+defined in the hardware map and subsequently flashing and running the tests
+on those platforms.
+
+.. note::
+
+  Currently only boards with support for both pyocd and nrfjprog are supported
```
Comment (dbkinder) [inline doc/guides/test/sanitycheck.rst]: Do the boards need to support **both pyocd and nrfjprog** ? I'm guessing it's either, and if so, change this to:
```
Currently only boards with support for either pyocd or nrfjprog are supported ...
```

**Review by dbkinder — CHANGES_REQUESTED** [review]

**Comment by galak** [comment]
Will open a new PR with the code now that its cleaned up.

**Pull request closed without merging**